File: packages/go-client/deletion.go
package consentshield

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/url"
)

// DeletionReason values mirror the server contract.
const (
	DeletionReasonConsentRevoked  = "consent_revoked"
	DeletionReasonRightsRequest   = "rights_request"
	DeletionReasonAccountClosed   = "account_closed"
	DeletionReasonRetentionExpiry = "retention_expiry"
	DeletionReasonOperatorAction  = "operator_action"
)

// TriggerDeletionParams are the inputs to TriggerDeletion.
type TriggerDeletionParams struct {
	PropertyID              string
	DataPrincipalIdentifier string
	IdentifierType          string
	Reason                  string
	PurposeCodes            []string
	ScopeOverride           string
	ActorType               string
	ActorRef                string
	ClientRequestID         string
	TraceID                 string
}

// TriggerDeletion enqueues a deletion across the customer's wired
// connectors. PurposeCodes is REQUIRED when Reason="consent_revoked"
// (compliance gate, fires before network).
func (c *Client) TriggerDeletion(ctx context.Context, p TriggerDeletionParams) (*DeletionTriggerEnvelope, error) {
	if err := requireStr(p.PropertyID, "PropertyID"); err != nil {
		return nil, err
	}
	if err := requireStr(p.DataPrincipalIdentifier, "DataPrincipalIdentifier"); err != nil {
		return nil, err
	}
	if err := requireStr(p.IdentifierType, "IdentifierType"); err != nil {
		return nil, err
	}
	if err := requireStr(p.Reason, "Reason"); err != nil {
		return nil, err
	}
	if p.Reason == DeletionReasonConsentRevoked && len(p.PurposeCodes) == 0 {
		return nil, errors.New("consentshield: PurposeCodes is required when Reason=consent_revoked")
	}
	for i, code := range p.PurposeCodes {
		if code == "" {
			return nil, fmt.Errorf("consentshield: PurposeCodes[%d] must be a non-empty string", i)
		}
	}

	body := map[string]any{
		"property_id":               p.PropertyID,
		"data_principal_identifier": p.DataPrincipalIdentifier,
		"identifier_type":           p.IdentifierType,
		"reason":                    p.Reason,
	}
	if len(p.PurposeCodes) > 0 {
		body["purpose_codes"] = p.PurposeCodes
	}
	if p.ScopeOverride != "" {
		body["scope_override"] = p.ScopeOverride
	}
	if p.ActorType != "" {
		body["actor_type"] = p.ActorType
	}
	if p.ActorRef != "" {
		body["actor_ref"] = p.ActorRef
	}
	if p.ClientRequestID != "" {
		body["client_request_id"] = p.ClientRequestID
	}

	resp, err := c.transport.do(ctx, &httpRequest{
		Method:  "POST",
		Path:    "/v1/deletion/trigger",
		Body:    body,
		TraceID: p.TraceID,
	})
	if err != nil {
		return nil, err
	}
	var out DeletionTriggerEnvelope
	if err := json.Unmarshal(resp.Body, &out); err != nil {
		return nil, fmt.Errorf("consentshield: decode trigger_deletion: %w", err)
	}
	return &out, nil
}

// ListDeletionReceiptsParams filters the receipts list.
type ListDeletionReceiptsParams struct {
	DeletionRequestID string
	Cursor            string
	Limit             int
	TraceID           string
}

func (p ListDeletionReceiptsParams) query() url.Values {
	q := url.Values{}
	if p.DeletionRequestID != "" {
		q.Set("deletion_request_id", p.DeletionRequestID)
	}
	if p.Cursor != "" {
		q.Set("cursor", p.Cursor)
	}
	if p.Limit > 0 {
		q.Set("limit", fmt.Sprintf("%d", p.Limit))
	}
	return q
}

// ListDeletionReceipts returns a single page of deletion receipts.
func (c *Client) ListDeletionReceipts(ctx context.Context, p ListDeletionReceiptsParams) (*DeletionReceiptsEnvelope, error) {
	resp, err := c.transport.do(ctx, &httpRequest{
		Method:  "GET",
		Path:    "/v1/deletion/receipts",
		Query:   p.query(),
		TraceID: p.TraceID,
	})
	if err != nil {
		return nil, err
	}
	var out DeletionReceiptsEnvelope
	if err := json.Unmarshal(resp.Body, &out); err != nil {
		return nil, fmt.Errorf("consentshield: decode list_deletion_receipts: %w", err)
	}
	return &out, nil
}

// DeletionReceiptPaginator walks every page of /v1/deletion/receipts.
type DeletionReceiptPaginator struct {
	client *Client
	params ListDeletionReceiptsParams
	done   bool
	last   []DeletionReceiptRow
	err    error
}

// IterateDeletionReceipts returns a paginator.
func (c *Client) IterateDeletionReceipts(p ListDeletionReceiptsParams) *DeletionReceiptPaginator {
	return &DeletionReceiptPaginator{client: c, params: p}
}

func (it *DeletionReceiptPaginator) Next(ctx context.Context) bool {
	if it.done {
		return false
	}
	env, err := it.client.ListDeletionReceipts(ctx, it.params)
	if err != nil {
		it.err = err
		it.done = true
		return false
	}
	it.last = env.Receipts
	if env.NextCursor == nil || *env.NextCursor == "" {
		it.done = true
	} else {
		it.params.Cursor = *env.NextCursor
	}
	return true
}

func (it *DeletionReceiptPaginator) Page() []DeletionReceiptRow { return it.last }
func (it *DeletionReceiptPaginator) Err() error                 { return it.err }

File: packages/node-client/src/deletion.ts
// ADR-1006 Phase 1 Sprint 1.3 — triggerDeletion + listDeletionReceipts.

import type { HttpClient } from './http'
import type {
  DeletionReason,
  DeletionReceiptRow,
  DeletionReceiptsEnvelope,
  DeletionTriggerEnvelope,
  IdentifierType,
} from './types'

const ALLOWED_REASONS: readonly DeletionReason[] = [
  'consent_revoked',
  'erasure_request',
  'retention_expired',
]

export type DeletionActorType = 'user' | 'operator' | 'system'

export interface TriggerDeletionInput {
  propertyId: string
  dataPrincipalIdentifier: string
  identifierType: IdentifierType | string
  reason: DeletionReason
  /** Required when reason === 'consent_revoked'; otherwise optional. */
  purposeCodes?: string[]
  /** Manual scope override — list of artefact ids to scope the delete to. */
  scopeOverride?: string[]
  actorType?: DeletionActorType
  actorRef?: string
  traceId?: string
  signal?: AbortSignal
}

export async function triggerDeletion(
  http: HttpClient,
  input: TriggerDeletionInput,
): Promise<DeletionTriggerEnvelope> {
  validateRequired(input.propertyId, 'propertyId')
  validateRequired(input.dataPrincipalIdentifier, 'dataPrincipalIdentifier')
  validateRequired(input.identifierType, 'identifierType')
  if (!ALLOWED_REASONS.includes(input.reason)) {
    throw new TypeError(
      `@consentshield/node: triggerDeletion input.reason must be one of: ${ALLOWED_REASONS.join(', ')}`,
    )
  }
  if (input.purposeCodes !== undefined) {
    if (!Array.isArray(input.purposeCodes)) {
      throw new TypeError('@consentshield/node: triggerDeletion input.purposeCodes must be an array')
    }
    for (let i = 0; i < input.purposeCodes.length; i++) {
      if (typeof input.purposeCodes[i] !== 'string' || input.purposeCodes[i].length === 0) {
        throw new TypeError(
          `@consentshield/node: triggerDeletion input.purposeCodes[${i}] must be a non-empty string`,
        )
      }
    }
  }
  if (input.reason === 'consent_revoked' && (!input.purposeCodes || input.purposeCodes.length === 0)) {
    throw new TypeError(
      "@consentshield/node: triggerDeletion input.purposeCodes is required when reason === 'consent_revoked'",
    )
  }
  if (input.scopeOverride !== undefined && !Array.isArray(input.scopeOverride)) {
    throw new TypeError('@consentshield/node: triggerDeletion input.scopeOverride must be an array')
  }
  if (input.actorType !== undefined && !['user', 'operator', 'system'].includes(input.actorType)) {
    throw new TypeError(
      "@consentshield/node: triggerDeletion input.actorType must be one of: user, operator, system",
    )
  }

  const body: Record<string, unknown> = {
    property_id: input.propertyId,
    data_principal_identifier: input.dataPrincipalIdentifier,
    identifier_type: input.identifierType,
    reason: input.reason,
  }
  if (input.purposeCodes !== undefined) body.purpose_codes = input.purposeCodes
  if (input.scopeOverride !== undefined) body.scope_override = input.scopeOverride
  if (input.actorType !== undefined) body.actor_type = input.actorType
  if (input.actorRef !== undefined) body.actor_ref = input.actorRef

  const resp = await http.request<DeletionTriggerEnvelope>({
    method: 'POST',
    path: '/deletion/trigger',
    body,
    signal: input.signal,
    traceId: input.traceId,
  })
  return resp.body
}

export interface ListDeletionReceiptsInput {
  triggerType?: string
  status?: string
  connectorId?: string
  createdAfter?: string
  createdBefore?: string
  cursor?: string
  limit?: number
  traceId?: string
  signal?: AbortSignal
}

export async function listDeletionReceipts(
  http: HttpClient,
  input: ListDeletionReceiptsInput = {},
): Promise<DeletionReceiptsEnvelope> {
  const resp = await http.request<DeletionReceiptsEnvelope>({
    method: 'GET',
    path: '/deletion/receipts',
    query: {
      trigger_type: input.triggerType,
      status: input.status,
      connector_id: input.connectorId,
      created_after: input.createdAfter,
      created_before: input.createdBefore,
      cursor: input.cursor,
      limit: input.limit,
    },
    signal: input.signal,
    traceId: input.traceId,
  })
  return resp.body
}

export async function* iterateDeletionReceipts(
  http: HttpClient,
  input: ListDeletionReceiptsInput = {},
): AsyncIterableIterator<DeletionReceiptRow> {
  let cursor = input.cursor
  while (true) {
    const page: DeletionReceiptsEnvelope = await listDeletionReceipts(http, { ...input, cursor })
    for (const item of page.items) yield item
    if (!page.next_cursor) return
    cursor = page.next_cursor
  }
}

function validateRequired(value: unknown, name: string): void {
  if (typeof value !== 'string' || value.length === 0) {
    throw new TypeError(`@consentshield/node: ${name} is required and must be a non-empty string`)
  }
}
